webrender/src/mask_cache.rs

```diff
@@ -99,59 +114,38 @@ impl MaskCacheInfo {
                   device_pixel_ratio: f32,
                   aux_lists: &AuxiliaryLists) {
-
         if self.local_rect.is_none() {
             let mut local_rect;
-            let mut local_inner: Option<LayerRect>;
             match source {
                 &ClipSource::NoClip => unreachable!(),
                 &ClipSource::Complex(rect, radius) => {
-                    let slice = clip_store.get_slice_mut(self.clip_range.start, CLIP_DATA_GPU_SIZE);
                     let data = ClipData::uniform(rect, radius);
-                    PrimitiveStore::populate_clip_data(slice, data);
-                    debug_assert_eq!(self.clip_range.item_count, 1);
+                    for (corner, component) in data.corners.iter().zip(self.corner_components.iter()) {
+                        let gpu_block = clip_store.get_mut(component.gpu_address);
+                        *gpu_block = GpuBlock32::from(*corner);
+                    }
                     local_rect = Some(rect);
-                    local_inner = ComplexClipRegion::new(rect, BorderRadius::uniform(radius))
-                                                    .get_inner_rect();
                 }
                 &ClipSource::Region(ref region) => {
                     local_rect = Some(LayerRect::from_untyped(&rect_from_points_f(-MAX_COORD, -MAX_COORD, MAX_COORD, MAX_COORD)));
-                    local_inner = match region.image_mask {
-                        Some(ref mask) if !mask.repeat => {
-                            local_rect = local_rect.and_then(|r| r.intersection(&mask.rect));
-                            None
-                        },
-                        Some(_) => None,
-                        None => local_rect,
-                    };
                     let clips = aux_lists.complex_clip_regions(&region.complex);
-                    assert_eq!(self.clip_range.item_count, clips.len() as u32);
-                    let slice = clip_store.get_slice_mut(self.clip_range.start, CLIP_DATA_GPU_SIZE * clips.len());
-                    for (clip, chunk) in clips.iter().zip(slice.chunks_mut(CLIP_DATA_GPU_SIZE)) {
+                    debug_assert_eq!(self.corner_components.len(), clips.len() * CORNERS_PER_CLIP_REGION);
+                    for (clip, chunk) in clips.iter().zip(self.corner_components.chunks_mut(CORNERS_PER_CLIP_REGION)) {
                         let data = ClipData::from_clip_region(clip);
-                        PrimitiveStore::populate_clip_data(chunk, data);
+                        for (corner, component) in data.corners.iter().zip(chunk) {
+                            let gpu_block = clip_store.get_mut(component.gpu_address);
+                            *gpu_block = GpuBlock32::from(*corner);
+                        }
                         local_rect = local_rect.and_then(|r| r.intersection(&clip.rect));
-                        local_inner = local_inner.and_then(|r| clip.get_inner_rect()
```

Comment:
AFAIK, our earlier discussion was about having 4 clip rectangles generated instead of one so that the inner area gets excluded from the mask computation. Then we figured that these 4 clips can be just the corners of a rectangle *if* the rounded rectangle is all we got. So for that model, we'd still need the inner area calculated. However, I'm not sure you need it for your implementation. I'll do another review pass with that in mind.